azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+    module=this_module,
+    all_ops=all_common_ops,
+    handled_ops=[
+        pipeline_ops_base.ConnectOperation,
+        pipeline_ops_base.DisconnectOperation,
+        pipeline_ops_base.ReconnectOperation,
+    ],
+    all_events=all_common_events,
+    handled_events=[],
+    extra_initializer_defaults={"blocked": False, "queue": queue.Queue},
+)
+
+connect_ops = [
+    {"op_class": pipeline_ops_base.ConnectOperation, "connected_flag_required_to_run": False},
+    {"op_class": pipeline_ops_base.DisconnectOperation, "connected_flag_required_to_run": True},
+    {"op_class": pipeline_ops_base.ReconnectOperation, "connected_flag_required_to_run": True},
+]
+
+
+class FakeOperation(pipeline_ops_base.PipelineOperation):
+    pass
+
+
+@pytest.mark.describe("SerializeConnectOpsStage - .run_op() -- called with connect operation")
+class TestSerializeConnectOpStageRunOp(object):
+    @pytest.fixture
+    def stage(self, mocker):
+        stage = make_mock_stage(
+            mocker=mocker, stage_to_make=pipeline_stages_base.SerializeConnectOpsStage
+        )
+        stage.next.run_op = mocker.MagicMock()
+        return stage
+
+    @pytest.fixture
+    def connect_op(self, mocker, params):
+        return params["op_class"](callback=mocker.MagicMock())
+
+    @pytest.fixture
+    def fake_op(self, mocker):
+        return FakeOperation(callback=mocker.MagicMock())
+
+    @pytest.fixture
+    def fake_ops(self, mocker):
+        return [
+            FakeOperation(callback=mocker.MagicMock()),
+            FakeOperation(callback=mocker.MagicMock()),
+            FakeOperation(callback=mocker.MagicMock()),
+        ]
+
+    @pytest.mark.it(
+        "Immediately completes a ConnectOperation if the transport is already connected"
+    )
+    def test_connect_while_connected(self, stage, mocker):
+        op = pipeline_ops_base.ConnectOperation(callback=mocker.MagicMock())
+        stage.pipeline_root.connected = True
+        stage.run_op(op)
+        assert_callback_succeeded(op=op)
+
+    @pytest.mark.it(
+        "Immediately completes a DisconnectOperation if the transport is already disconnected"
+    )
+    def test_disconnect_while_disconnected(self, stage, mocker):
+        op = pipeline_ops_base.DisconnectOperation(callback=mocker.MagicMock())
+        stage.pipeline_root.connected = False
+        stage.run_op(op)
+        assert_callback_succeeded(op=op)
+
+    @pytest.mark.it(
+        "Immediately passes the operation down if an operation is not alrady being serialized"
+    )
+    def test_passes_op_when_not_blocked(self, stage, mocker, fake_op):
+        stage.run_op(fake_op)
+        assert stage.next.run_op.call_count == 1
+        assert stage.next.run_op.call_args[0][0] == fake_op
+
+    @pytest.mark.parametrize(
+        "params", connect_ops, ids=[x["op_class"].__name__ for x in connect_ops]
+    )
+    @pytest.mark.it(
+        "Does not pass the operation down if a different operation is currently being serialized"
```

Comment:
"Does not immediately pass the operation down", since, it will, of course, eventually get passed down.

Arguably this, and the next test could be folded together for a greater degree of clarity.